### PR TITLE
Exponential changes

### PR DESCRIPTION
**Pull request by Pun-it**
Screenshot from the local testing.

image
image
image

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary
The PR adds a math‑string formatting utility (`formatSuperscript`) and integrates it across DOCX export and Angular UI rendering. The logic is straightforward, but there are several **severity‑HIGH** and **severity‑MEDIUM** issues around missing input typing, error‑handling violations, logging violations, and architecture violations (magic values, untyped DTOs). No critical issues, but several must be fixed.

---

## Issues

### 1. HIGH — Missing error handling & logging in **formatSuperscript**  
**File:** `math-formatting.util.ts`  
**Function:** `formatSuperscript`  
- Violates *Global ERROR HANDLING rules 1 & 3*.  
- If `text` is non‑string (e.g., number, null from API), `.replace` will throw without logs or fallback.  
**Fix:**  
Add a type‑check and log a structured warning (`WARNING` level) via the central logging utility, then return the unmodified value.

---

### 2. HIGH — Violates STRICT TYPING (uses `any`)  
**File:** `question-bank-download.service.ts`  
**Function:** internal loops where `q` and `opt` are declared (`forEach`)  
- `q: any` and `opt: any` violate STRICT TYPING rule 1.  
- DTOs for question and option should be typed since they are part of service‑level business logic.  
**Fix:**  
Create `QuestionDTO` and `OptionDTO` types in the shared models folder and apply them.

---

### 3. HIGH — HTML template calls `formatSuperscript(option)` where `option` may be an object  
**File:** `question-bank-view.component.html`  
**Template block rendering options**  
- If `option.text` is falsy but `option` is an object (common in mixed datasets), `formatSuperscript(option)` will receive an object → runtime error.  
**Fix:**  
Ensure `option` is coerced to string before passing to `formatSuperscript`, or refine the `*ngIf` conditions ensuring objects never fall into the non‑text path.

---

### 4. MEDIUM — No unit tests for new service-level logic  
**File:** `math-formatting.util.ts` and any usage in services  
Violates TESTING rule 1.  
**Fix:**  
Add unit tests covering at minimum: multi‑digit exponents, symbols, invalid input, null/undefined, and no-op.

---

### 5. MEDIUM — Hardcoded SUPERSCRIPT_MAP violates config centralisation  
**File:** `math-formatting.util.ts`  
Hardcoded map is acceptable for utility, but per ARCHITECTURE rule 1, constants must live in a central constants file.  
**Fix:**  
Move `SUPERSCRIPT_MAP` into a shared constants module.

---

### 6. LOW — Inconsistent typing in template (implicit `any[]`)  
**File:** `question-bank-view.component.ts`  
`shuffledColumns: any[] = [];` and `primaryColumn: any[] = [];` violate STRICT TYPING rule 1.  
**Fix:**  
Define proper typed arrays, e.g., `string[]` or a DTO.

---

### 7. LOW — Missing structured logs in DOCX generation paths  
**File:** `question-bank-download.service.ts`  
**Function:** main generation method using `formatSuperscript`  
- If `formatSuperscript` fails or provides malformed content, no logging exists.  
**Fix:**  
Add structured logs at start/end of export and around error boundaries.

---

## Score Calculation
- HIGH × 3 = -15  
- MEDIUM × 2 = -6  
- LOW × 2 = -2  

**Final Score = 100 − 23 = 77**

---

## (Optional) References
- Unicode superscript behaviour: https://www.unicode.org/charts/  
- Angular template type‑safety notes: https://angular.io/guide/template-typecheck

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  '+': '⁺', '-': '⁻', '(': '⁽', ')': '⁾',
+};
+
+/**
+ * Converts caret-based exponent notation (e.g. `2^3`, `x^10`) to
+ * Unicode superscript characters (e.g. `2³`, `x¹⁰`).
+ */
+export function formatSuperscript(text: string): string {
+  if (!text) return text;
+  return text.replace(/\^([0-9nxy()+\-]+)/g, (_, exponent: string) => {
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: What about other variable text like a^5 and z^10? The current regex seems to be configured to work for x,n,y only.

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts [hunk 2]
```diff
@@ -112,14 +113,14 @@ export class QuestionBankDownloadService {
       section.questions.forEach((q: any, index: number) => {
         if (q.question) {
           content.push(new Paragraph({
-            text: `${index + 1}. ${q.question}`,
+            text: `${index + 1}. ${formatSuperscript(q.question)}`,
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: While Word document can use Unicode characters, that is not the same as true DOCX superscript. Please validate a question paper download and resolve only if it is workable, else fix

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  '5': '⁵', '6': '⁶', '7': '⁷', '8': '⁸', '9': '⁹',
+  'n': 'ⁿ', 'x': 'ˣ', 'y': 'ʸ',
+  '+': '⁺', '-': '⁻', '(': '⁽', ')': '⁾',
+};
+
+/**
+ * Converts caret-based exponent notation (e.g. `2^3`, `x^10`) to
+ * Unicode superscript characters (e.g. `2³`, `x¹⁰`).
+ */
+export function formatSuperscript(text: string): string {
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: Please add unit tests here. **This was also highlighted by a4i-architect no. 4 in [comment](https://github.com/A4i-tech/Shiksha-Copilot/pull/65#pullrequestreview-4081984494)**

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html [hunk 4]
```diff
@@ -101,8 +101,8 @@ <h2 class="font-bold text-content">
                 class="list-alpha">
                 <li *ngFor="let option of questionData.options">
                   <!-- Options may be object entries ({ text, label? }) or plain strings. -->
-                  <span *ngIf="option.text">{{ option.text }}</span>
-                  <span *ngIf="!option.text">{{ option }}</span>
+                  <span *ngIf="option.text">{{ formatSuperscript(option.text) }}</span>
+                  <span *ngIf="!option.text">{{ formatSuperscript(option) }}</span>
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html]: We should pass only string into the format superscript, also check for string before passing it through. **This was also highlighted by a4i-architect no. 3 in [comment](https://github.com/A4i-tech/Shiksha-Copilot/pull/65#pullrequestreview-4081984494)**

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts (hunk 1 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: I've updated the regex to `/\^([a-zA-Z0-9()+\-]+)/g` and expanded the mapping dictionary to support the full `[a-z]` alphabet. Any alphabet characters passed that omit a standard Unicode superscript equivalent will fall back to default rendering without breaking.

**Review by a4i-architect — COMMENTED** [review]
## High‑Level Summary  
The PR introduces math superscript formatting across the UI and DOCX export. The logic is mostly correct, but a few important issues exist around typing, error‑handling expectations, and missing test coverage for newly added service behaviour.  

Below are **only issues on modified or added lines**, prioritized by severity.

---

## Issues

### 1. **Missing unit tests for new service‑level function**  
**File:** `question-bank-download.service.ts`  
**Function:** `convertToDocxRuns`  
**Severity:** **MEDIUM** (Testing requirement)  
**Issue:** The global testing standard requires at least one test for all new service‑level functions. `convertToDocxRuns` is new and untested.  
**Fix:** Add a dedicated spec file validating superscript detection, text splitting, and fallback behaviour.

---

### 2. **Typing violation: use of `any`**  
**File:** `question-bank-download.service.ts`  
**Functions:** modified loops in question/option processing  
**Severity:** **LOW**  
**Issue:** `q: any` and `opt: any` break the strict typing guideline.  
**Fix:** Replace `any` with an explicit DTO type, e.g., `QuestionDto` and `OptionDto`.

---

### 3. **Error-handling gap: convertToDocxRuns returns raw string for non-string input**  
**File:** `question-bank-download.service.ts`  
**Function:** `convertToDocxRuns`  
**Severity:** **MEDIUM**  
**Issue:** When text is not a string, the function creates a TextRun with the raw stringified input but does **not log** the irregular input, violating global error-handling rules (must log context for unexpected input).  
**Fix:** Add structured logging at WARNING level summarizing the input type before fallback.

---

### 4. **Missing fallback sanitisation in DOCX superScript rendering**  
**File:** `question-bank-download.service.ts`  
**Function:** `convertToDocxRuns`  
**Severity:** **LOW**  
**Issue:** The regex allows characters such as `+`, `-`, `(`, `)`, but rejects unsupported superscript characters (e.g., spaces or special math symbols). The resulting run may contain normal text that gets marked as superscript unexpectedly.  
**Fix:** Before creating a superscript TextRun, validate characters using SUPERSCRIPT_MAP and fallback to normal runs where mapping is absent.

---

### 5. **Magic numbers introduced in new DOCX rendering code**  
**File:** `question-bank-download.service.ts`  
**Functions:** modified TableRow paragraph creation  
**Severity:** **LOW**  
**Issue:** Added leading space `" "` and spacing parameters use inline literals, violating config centralisation rules.  
**Fix:** Move constants like left padding and spacing values to a shared config.

---

### 6. **Angular template: repeated heavy computation**  
**File:** `question-bank-view.component.html`  
**Lines:** new uses of `formatSuperscript(...)`  
**Severity:** **LOW**  
**Issue:** Calling a pure JS function repeatedly inside template loops triggers recalculation on every CD cycle.  
**Fix:** Wrap `formatSuperscript` in a pure pipe or precompute in `ngOnChanges`.

---

### 7. **Typing leak in component: exposing util function directly**  
**File:** `question-bank-view.component.ts`  
**Line:** `formatSuperscript = formatSuperscript;`  
**Severity:** **LOW**  
**Issue:** Exposing functions directly on the component instance encourages misuse and breaks testability.  
**Fix:** Create a proper wrapper method with type annotations.

---

## Score Calculation  
- MEDIUM: 2 issues × -2 = -4  
- LOW: 5 issues × -1 = -5  
**Total deduction: 9**

**Final Score = 100 - 9 = 91**

---

## Optional References  
- [Angular Performance Guide – Avoid heavy template computations](https://angular.io/guide/performance)

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts (hunk 2 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: refactored the download process inside `question-bank-download.service.ts` by adding a dedicated `convertToDocxRuns()` utility. Instead of formatting to Unicode string variants, this correctly interprets the caret syntax and applies the `docx` framework's `TextRun({ superScript: true })` capabilities natively, now Word yields true mathematical superscripts without artifacting.

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts (hunk 3 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: added tests in `math-formatting.util.spec.ts` covering exponent letters, parenthetical logic, empty/null parameters, unmapped fallback cases, and type safety constraints.

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html (hunk 4 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html]: Rather than repetitively making upstream checks, I've hardened the logic directly inside the `formatSuperscript` utility itself. It's now safely intercepting `null`/`undefined` triggers and asserting string types before mutating avoiding template-side runtime crashing.

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts [hunk 5]
```diff
@@ -148,8 +153,14 @@ export class QuestionBankDownloadService {
     for (let i = 0; i < col1.length; i++) {
       rows.push(new TableRow({
         children: [
-          new TableCell({ width: { size: 50, type: WidthType.PERCENTAGE }, children: [new Paragraph({text: ` ${formatSuperscript(col1[i])}`,spacing:{before:50, after:50}})] }),
-          new TableCell({ width: { size: 50, type: WidthType.PERCENTAGE }, children: [new Paragraph({text:` ${formatSuperscript(col2[i])}`,spacing:{before:50, after:50}})] }),
+          new TableCell({ width: { size: 50, type: WidthType.PERCENTAGE }, children: [new Paragraph({
+            children: [new TextRun({ text: " " }), ...this.convertToDocxRuns(col1[i])],
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: Why empty spaced text " "? We should directly allow function call replacements itself.

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts [hunk 6]
```diff
@@ -148,8 +153,14 @@ export class QuestionBankDownloadService {
...
+            spacing:{before:50, after:50}
+          })] }),
+          new TableCell({ width: { size: 50, type: WidthType.PERCENTAGE }, children: [new Paragraph({
+            children: [new TextRun({ text: " " }), ...this.convertToDocxRuns(col2[i])],
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: Why empty spaced text " "? We should directly allow function call replacements itself.

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  'p': 'ᵖ', 'r': 'ʳ', 's': 'ˢ', 't': 'ᵗ', 'u': 'ᵘ',
+  'v': 'ᵛ', 'w': 'ʷ', 'x': 'ˣ', 'y': 'ʸ', 'z': 'ᶻ',
+  '+': '⁺', '-': '⁻', '(': '⁽', ')': '⁾',
+};
+
+/**
+ * Converts caret-based exponent notation (e.g. `2^3`, `x^10`) to
+ * Unicode superscript characters (e.g. `2³`, `x¹⁰`).
+ */
+export function formatSuperscript(text: any): string {
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: We have introduced any again :/ . Please change to accept only specific types here. This is also highlighted by a4i-architect's [**comment in point 2**](https://github.com/A4i-tech/Shiksha-Copilot/pull/65#pullrequestreview-4117754629)

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html (hunk 4 above)
Comment (Soumabha) [reply shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html]: This is not the pattern we wish to follow. We have introduced `any` again in the function causing even more confusion. We want to ensure that all upstream do not pass objects and clarified strings only - if required sanitise the "options" coming from the backend to ensure a clean implementation. This is also highlighted by a4i-architect's [**comment in point 2**](https://github.com/A4i-tech/Shiksha-Copilot/pull/65#pullrequestreview-4117754629)

**Review by Soumabha — CHANGES_REQUESTED** [review]

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts [hunk 8]
```diff
@@ -160,4 +172,27 @@ export class QuestionBankDownloadService {
       ],
     });
   }
+
+  private convertToDocxRuns(text: string): TextRun[] {
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: Missing unit tests for the new function added. This is also highlighted by a4i-architect's [**comment in point 3**](https://github.com/A4i-tech/Shiksha-Copilot/pull/65#pullrequestreview-4117754629)

**Review by a4i-architect — COMMENTED** [review]
Below is your complete code review following all rules.

---

## High‑Level Summary

Overall the PR introduces superscript formatting, a new pipe, DTO typing, DOCX formatting logic, and unit tests. The PR is mostly solid, but there are **several violations of global standards** (logging, strict typing, error handling), and **a few functional bugs inside `convertToDocxRuns`**.  
The most severe issues:  
- improper logging (`console.warn`) in multiple functions, violating structured logging rules  
- regex in `convertToDocxRuns` produces incorrect results for many exponent edge‑cases  
- DTOs using optional fields without defaults causes unsafe undefined behavior downstream  
- missing timeouts for external file save call is OK (not a remote API), so no issue  
- missing tests for new functions (some new code lacks tests)

Final score computed at bottom.

---

## Issues

### 1. **Improper logging using `console.warn`**  
File: `math-formatting.util.ts`, function `formatSuperscript`  
File: `question-bank-download.service.ts`, function `convertToDocxRuns`  
Severity: **MEDIUM (-2)**  
Problem: Using `console.warn` violates the global logging standard: structured logging only, no console methods, and no logging of input types directly (risk of leaking sensitive data when extended).  
Fix: Replace with shared structured logger (`LoggerService.warn(...)`) and mask actual data type or values as required.

---

### 2. **Regex in convertToDocxRuns fails to handle many legal caret patterns**  
File: `question-bank-download.service.ts`, function `convertToDocxRuns`  
Severity: **HIGH (-3)**  
Problem: The regex only matches single caret + alphanumerics or parentheses. It fails for:  
- nested parentheses  
- caret next to caret (`^^`)  
- braces `{}` which appear in some math notations  
- expressions ending with caret (`x^`)  
Fix: Replace regex with a parser that properly identifies superscript sequences or extend regex to capture more cases.

---

### 3. **SUPERSCRIPT_MAP not used in convertToDocxRuns**  
File: `question-bank-download.service.ts`, function `convertToDocxRuns`  
Severity: **HIGH (-3)**  
Problem: Function emits superscript runs with `superScript: true` but **ignores SUPERSCRIPT_MAP in actual formatted TextRun output**. The tests imply mapped Unicode superscripts should be used.  
Fix: Convert exponent characters to their mapped values before creating the superscript TextRun.

---

### 4. **DTOs allow optional fields without safe defaults**  
File: `question-bank.dto.ts`, all interfaces  
Severity: **MEDIUM (-2)**  
Problem: Many fields are optional but downstream code does not handle missing values (e.g., question.text, option.label).  
Fix: Add defaults or ensure downstream consumer applies safe null-coalescing.

---

### 5. **Missing unit tests for buildQuestions and buildMatchTable**  
File: `question-bank-download.service.ts`  
Severity: **MEDIUM (-2)**  
Problem: PR adds tests only for convertToDocxRuns, not for new branching logic in buildQuestions nor match table handling.  
Fix: Add tests validating DOCX structure for both MCQ and Match-the-following branches.

---

### 6. **StructuredClone usage introduces potential mismatch for non‑serializable values**  
File: `question-bank-download.service.ts`, function `buildQuestions`  
Severity: **LOW (-1)**  
Problem: structuredClone drops functions, Dates, and complex types silently. Unclear if section.questions.value2 always contains plain strings.  
Fix: Replace with slice() when cloning simple arrays.

---

### 7. **Pipe does not guard against non‑string values**  
File: `math-formatting.pipe.ts`, function `transform`  
Severity: **LOW (-1)**  
Problem: transform calls formatSuperscript without checking for null/undefined. Tests cover util but pipe should enforce type guarantees for UI stability.  
Fix: Add an explicit typeof check before transformation.

---

### 8. **Unstructured type usage in QuestionBankViewComponent HTML**  
File: `question-bank-view.component.html`  
Severity: **LOW (-1)**  
Problem: Inline logic relies on option.text existing without typing enforcement.  
Fix: Introduce template-safe typed getter or stricter DTO.

---

## Score Calculation

Issues:  
- MEDIUM ×3 = -6  
- HIGH ×2 = -6  
- LOW ×3 = -3  

Total deduction = 15  
Score = 100 - 15 = **85**

---

## Final Score: **85 / 100**

---

## References
- Angular structured logging patterns: https://angular.io/guide/logging  
- Superscript Unicode reference: https://unicode.org/charts/PDF/U2070.pdf

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts (hunk 8 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: Added question-bank-download.service.spec.ts with tests covering: plain text, caret notation splitting, multi-digit exponents, parenthesized exponents, null/undefined/numeric fallbacks, empty strings, unmapped character warnings, and consecutive carets.

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts (hunk 6 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts]: Removed. The new TextRun({ text: " " }) before each convertToDocxRuns() call in buildMatchTable has been dropped. Table cells now directly use children: this.convertToDocxRuns(col[i]).

File: shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts (hunk 7 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/shared/utility/math-formatting.util.ts]: formatSuperscript signature changed from (text: any) to (text: string). A runtime guard with console.warn is still in place for safety, but the compile-time contract now enforces string-only input. Existing spec updated with as any casts to continue testing the runtime fallback.

File: shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html (hunk 4 above)
Comment (Pun-it) [reply shiksha-website/shiksha-frontend/src/app/view/user/question-bank/question-bank-view/question-bank-view.component.html]: Created strict OptionDto, QuestionDto, and QuestionSectionDto interfaces in shared/models/question-bank.dto.ts. All any usages for questions and options in question-bank-download.service.ts have been replaced with these typed interfaces. Options are now typed as string | { text?: string; label?: string } instead of any.

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    });
+
+    it('should handle text with no exponents returning full text', () => {
+      const runs = service.convertToDocxRuns('simple text');
+      expect(runs.length).toBe(1);
+    });
+
+    it('should handle consecutive carets', () => {
+      const runs = service.convertToDocxRuns('a^2b^3');
+      // 'a' + sup '2' + 'b' + sup '3'
```
Comment (Soumabha) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts:109]: It should be 'a' + sup '2b' + sup '3' and the runs should be 3, please double check the test impl and assertion.

**Review by Soumabha — APPROVED** [review]

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts [hunk 10]
```diff
@@ -102,53 +95,61 @@ export class QuestionBankDownloadService {
...
           }
-        }
-      });
-      } else if(section.type === 'Match the following'){
-          const colOneValue = section.questions.map((e:any)=> e.value1)
-          const colTwoVal = structuredClone(section.questions.map((e:any)=> e.value2))
-          const shuffedColums = this.utilityService.shuffleOptions(colTwoVal)
-          content.push(this.buildMatchTable(colOneValue, shuffedColums));
+        });
+      } else if (section.type === 'Match the following') {
```
Comment (mamuqsit) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts:125]: `else if` can be shortened to `else`, it is equivalent to the antithesis

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('convertToDocxRuns', () => {
+    it('should return a single TextRun for plain text without caret notation', () => {
+      const runs = service.convertToDocxRuns('Hello world');
+      expect(runs.length).toBe(1);
+    });
```
Comment (mamuqsit) [inline shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts:33]: Consider adding a test where you run an entire question text, something like:
```js
expect(service.convertToDocxRuns(
    "Identify the vertex of the parabola represented by the equation y = 2x^2 - 4x + 1."
)).toBe(
    "Identify the vertex of the parabola represented by the equation y = 2x² - 4x + 1."
);
```

Goal is to test the surrounding text remains unchanged. That is closer to real question-bank content than isolated tokens.

**Review by mamuqsit — CHANGES_REQUESTED** [review]

**Comment by mamuqsit** [comment]
Tested, functionality works in both - the preview UI as well as in the DOCX. Left minor review comments more on the testing and cleanup side.
image

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts (hunk 9 above)
Comment (mamuqsit) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts:109]: Addressed in https://github.com/A4i-tech/Shiksha-Copilot/pull/66/commits/e1cd6a849d3c3c185912a59a18eaeb1c1f1d259e

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts (hunk 10 above)
Comment (mamuqsit) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.ts:125]: Addressed in https://github.com/A4i-tech/Shiksha-Copilot/pull/66/commits/059b7e1485420ed0ec24660fea33af5a21370bdf

File: shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts (hunk 11 above)
Comment (mamuqsit) [reply shiksha-website/shiksha-frontend/src/app/shared/services/question-bank-download.service.spec.ts:33]: Addressed in https://github.com/A4i-tech/Shiksha-Copilot/pull/66/commits/56a87fff202605608ee82ed6f5289a30ae25e07b

**Review by Soumabha — APPROVED** [review]
